### PR TITLE
[STORM-2367] Documentation for streams API

### PR DESCRIPTION
**Pull request by arunmahadevan**

### CODE REVIEW

**Comment by HeartSaVioR** [comment]
+1 Nice explanation.

Since it's documentation you don't need to get +1 and even don't need to wait for 24hr. Or we can wait for more eyes. It's up to you.

**Comment by satishd** [comment]
@HeartSaVioR better to wait for 1-2 days to give a chance for others to review as this is not really urgent to push in short time to be part of some release. :) 
I will review this PR by today.

**Comment by HeartSaVioR** [comment]
Sure. :) Just would like to remind the bylaw for documentation.

File: docs/Stream-API.md [hunk 1]
```diff
@@ -0,0 +1,473 @@
...
+Stream<String> sentences = ...
+Stream<String> words = sentences.flatMap(s -> Arrays.asList(s.split(" ")));
+```
+
+
+In the above example, the lambda function splits each value in the stream to a list of words and the flatMap function generates a flattened stream of words out of it.
+
+## <a name="windowing"></a> Windowing
+
+A `window` operation produces a windowed stream consisting of the elements that fall within the window as specified by the window parameter. All the windowing options supported in the underlying windowed bolts is supported via the Stream apis.
```
Comment (satishd) [inline docs/Stream-API.md]: All the windowing options supported in the underlying windowed bolts ~~is~~ are supported via the Stream apis.

File: docs/Stream-API.md [hunk 2]
```diff
@@ -0,0 +1,473 @@
...
+
+```java
+// extract first three fields of the tuple emitted by the spout to produce a stream of typed tuples.
+Stream<Tuple3<String, Integer, Long>> stream = builder.newStream(new TestSpout(), TupleValueMappers.of(0, 1, 2));
+```
+
+# <a name="streamapis"></a> Stream APIs 
+
+Storm's streaming apis currently support a wide range of operations such as transformations, filters, windowing, aggregations, branching, joins, stateful, output and debugging operations.  
+
```
Comment (satishd) [inline docs/Stream-API.md]: You may want to give a link to Javadoc of `Stream` class.

File: docs/Stream-API.md [hunk 3]
```diff
@@ -0,0 +1,473 @@
...
+title: Stream API Overview
+layout: documentation
+documentation: true
+---
+
+* [Concepts](#concepts)
+    * [Stream Builder] (#streambuilder)
+    * [Value mapper] (#valuemapper)
+* [Stream APIs](#streamapis)
+    * [Basic transformations] (#basictransformations)
```
Comment (satishd) [inline docs/Stream-API.md]: Same as the above comment to remove space between [] and () for generating links till aggregations section.

File: docs/Stream-API.md [hunk 4]
```diff
@@ -0,0 +1,473 @@
+---
+title: Stream API Overview
+layout: documentation
+documentation: true
+---
+
+* [Concepts](#concepts)
+    * [Stream Builder] (#streambuilder)
```
Comment (arunmahadevan) [inline docs/Stream-API.md]: It seems the space does not matter. I am able to click on the links here - https://github.com/arunmahadevan/storm/blob/STORM-2367/docs/Stream-API.md

File: docs/Stream-API.md (hunk 3 above)
Comment (arunmahadevan) [inline docs/Stream-API.md]: It seems the space does not matter. I am able to click on the links here - https://github.com/arunmahadevan/storm/blob/STORM-2367/docs/Stream-API.md

File: docs/Stream-API.md (hunk 2 above)
Comment (arunmahadevan) [inline docs/Stream-API.md]: Added reference the the Stream/PairStream java files. Maybe the javadoc is not generated automatically,  I am not able to find the link.

File: docs/Stream-API.md (hunk 4 above)
Comment (satishd) [inline docs/Stream-API.md]: Some of the IDE MD plugins and tools are not generating links appropriately when there is a space. You can see the syntax [here](https://guides.github.com/features/mastering-markdown/) and [here] (https://en.support.wordpress.com/markdown-quick-reference/)

Even Github comments throws broken links when there is space between [] and (), you can observe above `here` links to see the difference. I guess it is better to address that.

File: docs/Stream-API.md (hunk 3 above)
Comment (satishd) [inline docs/Stream-API.md]: Some of the IDE MD plugins and tools are not generating links appropriately when there is a space. You can see the syntax [here](https://guides.github.com/features/mastering-markdown/) and [here] (https://en.support.wordpress.com/markdown-quick-reference/)

Even Github comments throws broken links when there is space between [] and (), you can observe above `here` links to see the difference. I guess it is better to address that.

Comment (arunmahadevan) [inline docs/Stream-API.md]: updated

**Comment by satishd** [comment]
+1, Nice summary of streams API. Thanks @arunmahadevan

**Pull request merged**